utils/db_utils.py

```diff
@@ -158,4 +159,87 @@ def add_user_stats(user_data):
             if last_call != -1:
                 user['last_call'] = arrow.get(last_call).format(time_format)
 
-    return user_data
\ No newline at end of file
+    return user_data
+
+def query_segments_crossing_endpoints(start_lat, start_long, end_lat, end_long, range_around_endpoints=400):
+    # data.loc only appears in analysis/recreated_location
+    query_start = {'data.loc': {'$near': SON([('$geometry', SON([('type', 'Point'), ('coordinates', [start_long, start_lat])])), ('$maxDistance', range_around_endpoints)])}}
+    query_end = {'data.loc': {'$near': SON([('$geometry', SON([('type', 'Point'), ('coordinates', [end_long, end_lat])])), ('$maxDistance', range_around_endpoints)])}}
+    
+    res_end = edb.get_analysis_timeseries_db().find(query_end)
+    end_by_section = {}
+    for elt in res_end:
+        elt_data = elt.get('data')
+        if elt_data:
+            section_id = elt_data.get('section')
+            # This makes sure that only the first encounter of the section is added
+            # The near query gives closest points first, so the first one is the one we want
+            if section_id not in end_by_section:
+                end_by_section[section_id] = elt
+    
```

Comment:
you can then use `get_data_df` with the geoquery from the timeseries, and this code can collapse to something like

```
agg_ts = eastag.AggregateTimeSeries().get_aggregate_time_series()
locs_matching_end = agg_ts.get_data_df("analysis/recreated_location", geo_query = gq)
end_by_section = locs_matching_end.section_id.unique()
```
and get the benefit of the C speedup

Reply:
Good point, I have rewritten the method using this. It isn't yet committed, since changes will be needed depending on the GeoQuery implementation.
I'm pasting the code here for future reference
```python
def query_segments_crossing_endpoints(start_lat, start_long, end_lat, end_long, range_around_endpoints=400):
    start_gq = GeoQuery(xxx, start_long, start_lat, range_around_endpoints) # TBD
    end_gq = GeoQuery(xxx, end_long, end_lat, range_around_endpoints) # TBD
    agg_ts = estag.AggregateTimeSeries().get_aggregate_time_series()

    locs_matching_start = agg_ts.get_data_df("analysis/recreated_location", geo_query = start_gq)
    locs_matching_start = locs_matching_start.drop_duplicates(subset=['section'])
    if locs_matching_start.empty:
        return locs_matching_start # No need to continue, no results
    
    locs_matching_end = agg_ts.get_data_df("analysis/recreated_location", geo_query = end_gq)
    locs_matching_end = locs_matching_end.drop_duplicates(subset=['section'])
    if locs_matching_end.empty:
        return locs_matching_end # No need to continue, no results
    
    merged = locs_matching_start.merge(locs_matching_end, how='outer', on=['section'])
    filtered = merged.loc[merged['idx_x']<merged['idx_y']].copy()
    filtered['duration'] = filtered['ts_y'] - filtered['ts_x']
    filtered['mode'] = filtered['mode_x']
    filtered['start_fmt_time'] = filtered['fmt_time_x']
    filtered['end_fmt_time'] = filtered['fmt_time_y']
    
    number_user_seen = filtered.user_id_x.nunique()

    if perm_utils.permissions.get("segment_trip_time_min_users", 0) <= number_user_seen:
        return filtered
    return pd.DataFrame.from_dict([])
```
Given the required database calls, I do not believe C speedup will make a big difference in the whole process, but eh, might as well try to optimize.